### PR TITLE
Feature/njv 372 implement quote feature in server

### PR DESCRIPTION
**Pull request by paykoob**
this is about beta quote implementation **just** on server-side 
there are still some tasks to do, e.g decide sending content of quotedMessage 
(I mark them with TODO hp)

@prettynatty this is for your review and comments.

### CODE REVIEW

File: actor-server/actor-bots/src/main/scala/im/actor/server/bot/BotToApiConversions.scala [hunk 1]
```diff
@@ -102,6 +102,7 @@ trait BotToApiConversions {
       case StickerMessage(stickerId, fastPreview, image512, image256, stickerCollectionId, stickerCollectionAccessHash) ⇒
         ApiStickerMessage(stickerId, fastPreview, image512, image256, stickerCollectionId, stickerCollectionAccessHash)
       case ServiceMessage(_)  ⇒ throw new RuntimeException("Service messages are not supported")
+      //      case EmptyMessage ⇒ ApiEmptyMessage // TODO needs pr
```
Comment (prettynatty) [inline actor-server/actor-bots/src/main/scala/im/actor/server/bot/BotToApiConversions.scala]: why?

File: actor-server/actor-core/src/main/scala/im/actor/api/rpc/HistoryImplicits.scala [hunk 2]
```diff
@@ -28,14 +29,21 @@ trait HistoryImplicits {
               Some(ApiMessageState.Sent)
           } else None // for incoming
 
+          //TODO hp: send content of message and ...
```
Comment (prettynatty) [inline actor-server/actor-core/src/main/scala/im/actor/api/rpc/HistoryImplicits.scala]: this TODO is open topic to be discussed

File: actor-server/actor-core/src/main/scala/im/actor/server/dialog/HistoryUtils.scala [hunk 3]
```diff
@@ -23,7 +23,9 @@ object HistoryUtils {
     date:                 DateTime,
     randomId:             Long,
     messageContentHeader: Int,
-    messageContentData:   Array[Byte]
+    messageContentData:   Array[Byte],
+    quotedMessagePeer:    Option[Peer],
+    quotedMessageRid:     Option[Long]
```
Comment (prettynatty) [inline actor-server/actor-core/src/main/scala/im/actor/server/dialog/HistoryUtils.scala]: we need to check that both of them are empty or non-empty

File: actor-server/actor-models/src/main/scala/im/actor/server/model/HistoryMessage.scala [hunk 4]
```diff
@@ -10,7 +10,9 @@ case class HistoryMessage(
   randomId:             Long,
   messageContentHeader: Int,
   messageContentData:   Array[Byte],
-  deletedAt:            Option[DateTime]
+  deletedAt:            Option[DateTime],
+  quotedMessagePeer:    Option[Peer]     = None,
+  quotedMessageRid:     Option[Long]     = None
```
Comment (prettynatty) [inline actor-server/actor-models/src/main/scala/im/actor/server/model/HistoryMessage.scala]: same here, check if noth of them empty or non-empty

File: actor-server/actor-persist/src/main/resources/sql/migration/V20160426151211__AddQuotedToHistory.sql [hunk 5]
```diff
@@ -0,0 +1,3 @@
+alter table history_messages add column quoted_peer_type int default 0;
+alter table history_messages add column quoted_peer_id int default 0;
+alter table history_messages add column quoted_random_id bigint default 0;
```
Comment (prettynatty) [inline actor-server/actor-persist/src/main/resources/sql/migration/V20160426151211__AddQuotedToHistory.sql]: why not nullable and default null?

File: actor-server/actor-rpc-api/src/main/scala/im/actor/server/api/rpc/service/messaging/MessagingHandlers.scala [hunk 6]
```diff
@@ -37,26 +38,44 @@ private[messaging] trait MessagingHandlers extends PeersImplicits
...
+            accessHash = Some(outPeer.accessHash),
+            forUserId = isOnlyForUser,
+            isFat = false,
+            quotedHistoryMessage = histMessage
+          ))
+        } yield ResponseSeqDate(s.seq, s.state.toByteArray, s.date)).value
+      }
+    }
+
+  def withQuoteMessage(quotedMessageReference: Option[ApiMessageOutReference])(f: Option[HistoryMessage] ⇒ Future[HandlerResult[ResponseSeqDate]]): Future[HandlerResult[ResponseSeqDate]] = {
```
Comment (prettynatty) [inline actor-server/actor-rpc-api/src/main/scala/im/actor/server/api/rpc/service/messaging/MessagingHandlers.scala]: private

File: actor-server/actor-core/src/main/scala/im/actor/server/dialog/DialogExtension.scala [hunk 7]
```diff
@@ -57,18 +58,24 @@ final class DialogExtensionImpl(system: ActorSystem) extends DialogExtension wit
...
+    message:              ApiMessage,
+    accessHash:           Option[Long]           = None,
+    isFat:                Boolean                = false,
+    forUserId:            Option[UserId]         = None,
+    quotedHistoryMessage: Option[HistoryMessage] = None
   ): Future[SeqStateDate] =
     withValidPeer(peer.asModel, senderUserId, Future.successful(SeqStateDate())) {
+      //todo hp: send content of message and public id?
       // we don't set date here, cause actual date set inside dialog processor
+      val quotedMessage = quotedHistoryMessage flatMap (hm ⇒ ApiMessage.parseFrom(hm.messageContentData).fold(
```
Comment (prettynatty) [inline actor-server/actor-core/src/main/scala/im/actor/server/dialog/DialogExtension.scala]: need to check if there is no empty message when we quote another message – we need a reply

**Comment by prettynatty** [comment]
Also, there are some more complex conditions of filling publicGroupId and quotedMessageId:
* you fill both if user forwards message to another chat and receiving user has access to the original group
* you don't fill quotedMessageId nor publicGroupId if user forwards message to another chat and receiving user doesn't has access to the original group (or original chat is private)
* you fill only quotedMessageId user replies to a message in the same chat

File: actor-server/actor-bots/src/main/scala/im/actor/server/bot/BotToApiConversions.scala (hunk 1 above)
Comment (paykoob) [inline actor-server/actor-bots/src/main/scala/im/actor/server/bot/BotToApiConversions.scala]: I added that, but nothing had changed, it seemed referenced code is in a external library:
actor-bots-shared-1.0.81 , but im not sure

File: actor-server/actor-persist/src/main/resources/sql/migration/V20160426151211__AddQuotedToHistory.sql (hunk 5 above)
Comment (paykoob) [inline actor-server/actor-persist/src/main/resources/sql/migration/V20160426151211__AddQuotedToHistory.sql]: you are right! I never comfort with sqls!

File: actor-server/actor-core/src/main/scala/im/actor/server/dialog/DialogExtension.scala (hunk 7 above)
Comment (paykoob) [inline actor-server/actor-core/src/main/scala/im/actor/server/dialog/DialogExtension.scala]: mm, I suggest one principle:
if bob reply to alice's message, bob call the SendMessage rpc with a QuotedMessageReference **with**  valid content (Text or Document ...).

bot if bob forwards alice's message  he just call SendMesage rpc with QuotedMessageReference and **EmptyMessage**
(and if bob has a comment on forwarded message, he should send a normal message by calling the sendMessage rpc (maybe before of forward) )

I mean
 Reply:
send message with valid content
Forward:
send message with empty content.

**Comment by paykoob** [comment]
ok, I consider the conditions, and update the PR.

**Comment by paykoob** [comment]
@prettynatty : whats about pubilcGroupId in new API. there is only channel and group.
should we send the publicGroupId when group is just channel?
@ex3ndr : can we have public/private group/channel? (maybe by using canInvitePeople property)

**Comment by prettynatty** [comment]
> whats about pubilcGroupId in new API. there is only channel and group.
should we send the publicGroupId when group is just channel?

yes

> can we have public/private group/channel? (maybe by using canInvitePeople property)

any group should be in publicGroupId

**Comment by paykoob** [comment]
@prettynatty I update the pr. after merging the new channel/group api we can consider it.

**Comment by paykoob** [comment]
maybe there is a bug on sending quotedMessageID! I should check it!

File: actor-server/actor-bots/src/main/scala/im/actor/server/bot/BotToApiConversions.scala (hunk 1 above)
Comment (prettynatty) [inline actor-server/actor-bots/src/main/scala/im/actor/server/bot/BotToApiConversions.scala]: upgrade actor-bots-shared to the latest version like it did in master, and see if code compiles

File: actor-server/actor-core/src/main/protobuf/dialog.proto [hunk 8]
```diff
@@ -239,6 +239,16 @@ message DialogCommands {
 
     message WriteMessageSelfAck {}
 
+    message QuotedMessage {
```
Comment (prettynatty) [inline actor-server/actor-core/src/main/protobuf/dialog.proto]: quoted message body should be determined inside DialogCommandHandlers, not in DialogExtension's api clients.

File: actor-server/actor-core/src/main/protobuf/dialog.proto [hunk 9]
```diff
@@ -252,6 +262,7 @@ message DialogCommands {
         google.protobuf.Int64Value access_hash = 10;
         bool is_fat = 6;
         google.protobuf.Int32Value for_user_id = 12;
+        QuotedMessage quoted_message = 13;
```
Comment (prettynatty) [inline actor-server/actor-core/src/main/protobuf/dialog.proto]: you need only reference here

File: actor-server/actor-core/src/main/protobuf/dialog.proto (hunk 8 above)
Comment (paykoob) [inline actor-server/actor-core/src/main/protobuf/dialog.proto]: do u mean I should just send apiMessageOutReference from MessagingHandlers to DialogExtension and then to DialogCommandHandlers?
thenDialogCommandHandlers build the body of quotedMessage?

I should add two parameter to SendMessageProto for holdin apiMessageOutReference. (to send them to DialogCommandHandlers)

Comment (prettynatty) [inline actor-server/actor-core/src/main/protobuf/dialog.proto]: Right.
Consider dialogExt as a lightweight interface to `DialogRoot` and `Dialog`. All business-logic contains inside this two actors.

**Pull request closed without merging**